src/libpanic_unwind/gcc.rs

```diff
@@ -106,117 +106,96 @@ fn rust_exception_class() -> uw::_Unwind_Exception_Class {
-                                exception_class: uw::_Unwind_Exception_Class,
-                                ue_header: *mut uw::_Unwind_Exception,
-                                context: *mut uw::_Unwind_Context)
-                                -> uw::_Unwind_Reason_Code;
-    }
+    // Register ids were lifted from LLVM's TargetLowering::getExceptionPointerRegister()
+    // and TargetLowering::getExceptionSelectorRegister() for each architecture,
+    // then mapped to DWARF register numbers via register definition tables
+    // (typically <arch>RegisterInfo.td, search for "DwarfRegNum").
+    // See also http://llvm.org/docs/WritingAnLLVMBackend.html#defining-a-register.
 
-    #[lang = "eh_personality"]
-    #[no_mangle]
-    extern "C" fn rust_eh_personality(version: c_int,
-                                      actions: uw::_Unwind_Action,
-                                      exception_class: uw::_Unwind_Exception_Class,
-                                      ue_header: *mut uw::_Unwind_Exception,
-                                      context: *mut uw::_Unwind_Context)
-                                      -> uw::_Unwind_Reason_Code {
-        unsafe { __gcc_personality_v0(version, actions, exception_class, ue_header, context) }
-    }
+    #[cfg(target_arch = "x86")]
+    const UNWIND_DATA_REG: (i32, i32) = (0, 2); // EAX, EDX
 
-    #[lang = "eh_personality_catch"]
-    #[no_mangle]
-    pub extern "C" fn rust_eh_personality_catch(version: c_int,
-                                                actions: uw::_Unwind_Action,
-                                                exception_class: uw::_Unwind_Exception_Class,
-                                                ue_header: *mut uw::_Unwind_Exception,
-                                                context: *mut uw::_Unwind_Context)
-                                                -> uw::_Unwind_Reason_Code {
+    #[cfg(target_arch = "x86_64")]
+    const UNWIND_DATA_REG: (i32, i32) = (0, 1); // RAX, RDX
 
-        if (actions as c_int & uw::_UA_SEARCH_PHASE as c_int) != 0 {
-            // search phase
-            uw::_URC_HANDLER_FOUND // catch!
-        } else {
-            // cleanup phase
-            unsafe { __gcc_personality_v0(version, actions, exception_class, ue_header, context) }
-        }
-    }
-}
+    #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+    const UNWIND_DATA_REG: (i32, i32) = (0, 1); // R0, R1 / X0, X1
 
-// iOS on armv7 is using SjLj exceptions and therefore requires to use
-// a specialized personality routine: __gcc_personality_sj0
+    #[cfg(any(target_arch = "mips", target_arch = "mipsel"))]
+    const UNWIND_DATA_REG: (i32, i32) = (4, 5); // A0, A1
 
-#[cfg(all(target_os = "ios", target_arch = "arm"))]
-pub mod eabi {
-    use unwind as uw;
-    use libc::c_int;
+    #[cfg(any(target_arch = "powerpc", target_arch = "powerpc64"))]
+    const UNWIND_DATA_REG: (i32, i32) = (3, 4); // R3, R4 / X3, X4
```

Comment:
Hm I couldn't seem to find these in LLVM's source, do you know where you found them?

```
lib/Target/PowerPC/PPCRegisterInfo.td:  def R#Index : GPR<Index, "r"#Index>, DwarfRegNum<[-2, Index]>;
lib/Target/PowerPC/PPCRegisterInfo.td:                    DwarfRegNum<[Index, -2]>;
lib/Target/PowerPC/PPCRegisterInfo.td:                DwarfRegNum<[!add(Index, 32), !add(Index, 32)]>;
lib/Target/PowerPC/PPCRegisterInfo.td:                 DwarfRegNum<[!add(Index, 32), !add(Index, 32)]>;
lib/Target/PowerPC/PPCRegisterInfo.td:                DwarfRegNum<[!add(Index, 77), !add(Index, 77)]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CR0 : CR<0, "cr0", [CR0LT, CR0GT, CR0EQ, CR0UN]>, DwarfRegNum<[68, 68]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CR1 : CR<1, "cr1", [CR1LT, CR1GT, CR1EQ, CR1UN]>, DwarfRegNum<[69, 69]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CR2 : CR<2, "cr2", [CR2LT, CR2GT, CR2EQ, CR2UN]>, DwarfRegNum<[70, 70]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CR3 : CR<3, "cr3", [CR3LT, CR3GT, CR3EQ, CR3UN]>, DwarfRegNum<[71, 71]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CR4 : CR<4, "cr4", [CR4LT, CR4GT, CR4EQ, CR4UN]>, DwarfRegNum<[72, 72]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CR5 : CR<5, "cr5", [CR5LT, CR5GT, CR5EQ, CR5UN]>, DwarfRegNum<[73, 73]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CR6 : CR<6, "cr6", [CR6LT, CR6GT, CR6EQ, CR6UN]>, DwarfRegNum<[74, 74]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CR7 : CR<7, "cr7", [CR7LT, CR7GT, CR7EQ, CR7UN]>, DwarfRegNum<[75, 75]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def LR  : SPR<8, "lr">, DwarfRegNum<[-2, 65]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def LR8 : SPR<8, "lr">, DwarfRegNum<[65, -2]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CTR  : SPR<9, "ctr">, DwarfRegNum<[-2, 66]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CTR8 : SPR<9, "ctr">, DwarfRegNum<[66, -2]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def VRSAVE: SPR<256, "vrsave">, DwarfRegNum<[109]>;
lib/Target/PowerPC/PPCRegisterInfo.td:def CARRY: SPR<1, "ca">, DwarfRegNum<[76]>;
```